dll/win32/browseui/CAutoComplete.cpp

```diff
@@ -25,664 +25,2124 @@
+    if (strText.IsEmpty())
+        return;
+
+    INT cItems = m_innerList.GetSize(); // get the number of items
+    if (cItems == 0)
+        return; // don't append
+
+    // get the common string
+    CStringW strCommon;
+    for (INT iItem = 0; iItem < cItems; ++iItem)
+    {
+        const CString& strItem = m_innerList[iItem]; // get the text of the item
+
+        if (iItem == 0) // the first item
+        {
+            strCommon = strItem; // store the text
+            continue;
+        }
+
+        for (INT ich = 0; ich < strCommon.GetLength(); ++ich)
+        {
+            if (ich < strItem.GetLength() && strCommon[ich] != strItem[ich])
+            {
+                strCommon = strCommon.Left(ich); // shrink the common string
+                break;
+            }
+        }
+    }
+
+    if (strCommon.IsEmpty() || strCommon.GetLength() <= strText.GetLength())
+        return; // no suggestion
+
+    // append suggestion
+    INT cchAppend = strCommon.GetLength() - strText.GetLength();
+    CStringW strAppend = strCommon.Right(cchAppend);
+    strText += strAppend;
+    SetEditText(strText);
+
+    // select the appended
+    INT ich0 = strText.GetLength();
+    INT ich1 = ich0 + cchAppend;
+    SetEditSel(ich0, ich1);
+}
+
+// go back a word ([Ctrl]+[Backspace])
+VOID CAutoComplete::DoBackWord()
+{
+    // get current selection
+    INT ich0, ich1;
+    m_hwndEdit.SendMessageW(EM_GETSEL,
+                            reinterpret_cast<WPARAM>(&ich0),
+                            reinterpret_cast<LPARAM>(&ich1));
+    if (ich0 <= 0 || ich0 != ich1)
+        return;
+    // get text
+    CStringW str = GetEditText();
+    // extend the range
+    while (ich0 > 0 && IsWordBreak(str[ich0 - 1]))
+        --ich0;
+    while (ich0 > 0 && !IsWordBreak(str[ich0 - 1]))
+        --ich0;
+    // select range
+    SetEditSel(ich0, ich1);
+    // replace selection with empty text (this is actually deletion)
+    m_hwndEdit.SendMessageW(EM_REPLACESEL, TRUE, (LPARAM)L"");
+}
+
+VOID CAutoComplete::UpdateScrollBar()
+{
+    // copy scroll info from m_hwndList to m_hwndScrollBar
+    SCROLLINFO si = { sizeof(si), SIF_ALL };
+    m_hwndList.GetScrollInfo(SB_VERT, &si);
+    m_hwndScrollBar.SetScrollInfo(SB_CTL, &si, FALSE);
+
+    // show/hide scroll bar
+    INT cVisibles = m_hwndList.GetVisibleCount();
+    INT cItems = m_hwndList.GetItemCount();
+    BOOL bShowScroll = (cItems > cVisibles);
+    m_hwndScrollBar.ShowWindow(bShowScroll ? SW_SHOWNOACTIVATE : SW_HIDE);
+    if (bShowScroll)
+        m_hwndScrollBar.InvalidateRect(NULL, FALSE); // redraw
+}
+
+BOOL CAutoComplete::OnEditKeyDown(WPARAM wParam, LPARAM lParam)
+{
+    TRACE("CAutoComplete::OnEditKeyDown(%p, %p)\n", this, wParam);
+
+    UINT vk = (UINT)wParam; // virtual key
+    switch (vk)
+    {
+        case VK_HOME: case VK_END: // [Home]/[End] key
+        case VK_UP: case VK_DOWN: // [Arrow Up]/[Arrow Down] key
+        case VK_PRIOR: case VK_NEXT: // [PageUp]/[PageDown] key
```

Comment:
No need for the comments that describe the keys, since this is well-known from the VK_ defines names.

Reply:
Deleted.